File: backend/app/schemas/canon.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, ConfigDict, Field


class CanonEntityOut(BaseModel):
    id: int
    reference_id: int
    entity_type: str
    canonical_name: str
    aliases: List[str] = Field(default_factory=list)
    summary: Optional[str] = None
    attributes: Dict[str, Any] = Field(default_factory=dict)
    source_refs: List[Dict[str, Any]] = Field(default_factory=list)
    importance: str
    confidence: float
    review_status: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CanonEntityCreate(BaseModel):
    entity_type: str
    canonical_name: str
    aliases: List[str] = Field(default_factory=list)
    summary: Optional[str] = None
    attributes: Dict[str, Any] = Field(default_factory=dict)
    source_refs: List[Dict[str, Any]] = Field(default_factory=list)
    importance: str = "major"


class CanonEntityUpdate(BaseModel):
    canonical_name: Optional[str] = None
    aliases: Optional[List[str]] = None
    summary: Optional[str] = None
    attributes: Optional[Dict[str, Any]] = None
    importance: Optional[str] = None
    review_status: Optional[str] = None


class CanonJobOut(BaseModel):
    id: int
    reference_id: int
    status: str
    model: Optional[str] = None
    chunk_total: int
    chunk_done: int
    failed_chunks: int
    entity_count: int
    error: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class CanonRelationOut(BaseModel):
    id: int
    reference_id: int
    source_entity_id: int
    target_entity_id: int
    relation_type: str
    label: Optional[str] = None
    summary: Optional[str] = None
    source_refs: List[Dict[str, Any]] = Field(default_factory=list)
    confidence: float
    review_status: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CanonRelationCreate(BaseModel):
    source_entity_id: int
    target_entity_id: int
    relation_type: str
    label: Optional[str] = None
    summary: Optional[str] = None
    source_refs: List[Dict[str, Any]] = Field(default_factory=list)


class CanonRelationUpdate(BaseModel):
    relation_type: Optional[str] = None
    label: Optional[str] = None
    summary: Optional[str] = None
    review_status: Optional[str] = None


class CanonGraphOut(BaseModel):
    nodes: List[CanonEntityOut] = Field(default_factory=list)
    edges: List[CanonRelationOut] = Field(default_factory=list)

File: frontend/src/api/canon.ts
import request from './request'

export type CanonEntityType =
  | 'character' | 'location' | 'ability' | 'faction' | 'worldrule'
  | 'event' | 'item' | 'race' | 'realm' | 'concept'

export type CanonImportance = 'critical' | 'major' | 'minor'

export type CanonReviewStatus =
  | 'ai_extracted'
  | 'user_verified'
  | 'user_edited'
  | 'user_added'

export type CanonJobStatus = 'pending' | 'processing' | 'done' | 'failed'

export interface CanonSourceRef {
  chapter?: number | string
  offset?: number
  quote?: string
}

export interface CanonEntity {
  id: number
  reference_id: number
  entity_type: CanonEntityType
  canonical_name: string
  aliases: string[]
  summary: string | null
  attributes: Record<string, unknown>
  source_refs: CanonSourceRef[]
  importance: CanonImportance
  confidence: number
  review_status: CanonReviewStatus
  created_at: string
  updated_at: string | null
}

export interface CanonJob {
  id: number
  reference_id: number
  status: CanonJobStatus
  model?: string | null
  chunk_total: number
  chunk_done: number
  failed_chunks: number
  entity_count: number
  error?: string | null
  created_at: string
  updated_at: string | null
}

export interface CanonEntityCreate {
  entity_type: CanonEntityType
  canonical_name: string
  aliases: string[]
  summary?: string | null
  attributes: Record<string, unknown>
  source_refs: CanonSourceRef[]
  importance?: CanonImportance
}

export interface CanonEntityUpdate {
  canonical_name?: string
  aliases?: string[]
  summary?: string | null
  attributes?: Record<string, unknown>
  importance?: CanonImportance
  review_status?: CanonReviewStatus
}

export type CanonRelationReview =
  | 'ai_extracted' | 'user_verified' | 'user_edited' | 'user_added'

export interface CanonRelation {
  id: number
  reference_id: number
  source_entity_id: number
  target_entity_id: number
  relation_type: string
  label: string | null
  summary: string | null
  source_refs: CanonSourceRef[]
  confidence: number
  review_status: CanonRelationReview
  created_at: string
  updated_at: string | null
}

export interface CanonGraph {
  nodes: CanonEntity[]
  edges: CanonRelation[]
}

export interface CanonRelationCreate {
  source_entity_id: number
  target_entity_id: number
  relation_type: string
  label?: string | null
  summary?: string | null
  source_refs?: CanonSourceRef[]
}

export interface CanonRelationUpdate {
  relation_type?: string
  label?: string | null
  summary?: string | null
  review_status?: CanonRelationReview
}

export const canonApi = {
  extract(refId: number) {
    return request.post(`/references/${refId}/canon/extract`, undefined, { skipErrorToast: true })
  },

  getJob(refId: number) {
    return request.get<CanonJob>(`/references/${refId}/canon/job`)
  },

  listEntities(refId: number) {
    return request.get<CanonEntity[]>(`/references/${refId}/canon/entities`)
  },

  createEntity(refId: number, data: CanonEntityCreate) {
    return request.post<CanonEntity>(`/references/${refId}/canon/entities`, data)
  },

  updateEntity(refId: number, entityId: number, data: CanonEntityUpdate) {
    return request.put<CanonEntity>(`/references/${refId}/canon/entities/${entityId}`, data)
  },

  deleteEntity(refId: number, entityId: number) {
    return request.delete(`/references/${refId}/canon/entities/${entityId}`)
  },

  getStreamTicket(refId: number) {
    return request.post<{ ticket: string }>(`/references/${refId}/canon/stream/ticket`)
  },

  getStreamUrl(refId: number, ticket: string): string {
    const base = (import.meta.env.VITE_API_BASE_URL || '').replace(/\/$/, '')
    return `${base}/api/v1/references/${refId}/canon/stream?ticket=${encodeURIComponent(ticket)}`
  },

  getGraph(refId: number) {
    return request.get<CanonGraph>(`/references/${refId}/canon/graph`)
  },

  listRelations(refId: number) {
    return request.get<CanonRelation[]>(`/references/${refId}/canon/relations`)
  },

  createRelation(refId: number, data: CanonRelationCreate) {
    return request.post<CanonRelation>(`/references/${refId}/canon/relations`, data)
  },

  updateRelation(refId: number, id: number, data: CanonRelationUpdate) {
    return request.put<CanonRelation>(`/references/${refId}/canon/relations/${id}`, data)
  },

  deleteRelation(refId: number, id: number) {
    return request.delete(`/references/${refId}/canon/relations/${id}`)
  },

  extractRelations(refId: number) {
    return request.post(`/references/${refId}/canon/extract-relations`, undefined, { skipErrorToast: true })
  },
}
